pinot-core/src/main/java/org/apache/pinot/core/data/manager/realtime/IngestionDelayTracker.java

```diff
@@ -0,0 +1,361 @@
+
+  private final boolean _enablePerPartitionMetric;
+  private final boolean _enableAggregateMetric;
+
+  private final RealtimeTableDataManager _realTimeTableDataManager;
+  private final Supplier<Boolean> _isServerReadyToServeQueries;
+
+  private Clock _clock;
+
+  /*
+   * Returns the maximum ingestion delay amongst all partitions we are tracking.
+   */
+  private DelayMeasure getMaximumDelay() {
+    DelayMeasure newMax = null;
+    for (int partitionGroupId : _partitionToDelaySampleMap.keySet()) {
+      DelayMeasure currentMeasure = _partitionToDelaySampleMap.get(partitionGroupId);
+      if ((newMax == null)
+          ||
+          (currentMeasure != null) && (currentMeasure._delayMs > newMax._delayMs)) {
+        newMax = currentMeasure;
+      }
+    }
+    return newMax;
+  }
+
+  /*
+   * Helper function to age a delay measure. Aging means adding the time elapsed since the measure was
+   * taken till the measure is being reported.
+  *
+  * @param currentDelay original sample delay to which we will add the age of the measure.
+   */
+  private long getAgedDelay(DelayMeasure currentDelay) {
+    if (currentDelay == null) {
+      return 0; // return 0 when not initialized
+    }
+    // Add age of measure to the reported value
+    long measureAgeInMs = _clock.millis() - currentDelay._sampleTime;
+    // Correct to zero for any time shifts due to NTP or time reset.
+    measureAgeInMs = Math.max(measureAgeInMs, 0);
+    return currentDelay._delayMs + measureAgeInMs;
+  }
+
+  /*
+   * Helper function to be called when we should stop tracking a given partition. Removes the partition from
+   * all our maps.
+   *
+   * @param partitionGroupId partition ID which we should stop tracking.
+   */
+  private void removePartitionId(int partitionGroupId) {
+    _partitionToDelaySampleMap.remove(partitionGroupId);
+    // If we are removing a partition we should stop reading its ideal state.
+    _partitionsMarkedForVerification.remove(partitionGroupId);
+    if (_enablePerPartitionMetric) {
+      _serverMetrics.removePartitionGauge(_metricName, partitionGroupId, ServerGauge.TABLE_INGESTION_DELAY_MS);
+    }
+  }
+
+  /*
+   * Helper functions that creates a list of all the partitions that are marked for verification and whose
+   * timeouts are expired. This helps us optimize checks of the ideal state.
+   */
+  private ArrayList<Integer> getPartitionsToBeVerified() {
+    ArrayList<Integer> partitionsToVerify = new ArrayList<>();
+    for (int partitionGroupId : _partitionsMarkedForVerification.keySet()) {
+      long markTime = _partitionsMarkedForVerification.get(partitionGroupId);
+      long timeMarked = _clock.millis() - markTime;
+      if (timeMarked > PARTITION_TIMEOUT_MS) {
+        // Partition must be verified
+        partitionsToVerify.add(partitionGroupId);
+      }
+    }
+    return partitionsToVerify;
+  }
+
+  public IngestionDelayTracker(ServerMetrics serverMetrics, String tableNameWithType,
```

Comment:
(minor) Put the constructors above other methods for readability

Reply:
done see commit 3ce7b783c5dc88073106f400f00783ba4ab71281